src/groups/mqb/mqbu/mqbu_resourceusagemonitor.h

```diff
@@ -658,6 +683,33 @@ inline ResourceUsageMonitorState::Enum ResourceUsageMonitor::state() const
     return bsl::max(d_bytes.state(), d_messages.state());
 }
 
+// --------------------
+// struct SingleCounter
+// --------------------
+
+inline SingleCounter::SingleCounter(SingleCounter* parent)
+: d_parent_p(parent)
+, d_value(0)
+{
+    // NOTHING
+}
+
+inline void SingleCounter::update(int delta)
+{
+    d_value += delta;
+
+    BSLS_ASSERT_SAFE(d_value >= 0);
+
+    if (d_parent_p) {
```

Comment:
Do we necessary need to have a tree-like structure for counting unconfirmed with update on every node on the path to the root?
I think, we only need one top-level counter during shutdown.
I would like to avoid `if` here, so we don't have extra branching in the normal workflow.

We can still build `SingleCounter`s with a parent object, but we can go up the tree and set up the pointer to the parent counter in each node. So we just update the atomic here and don't call `d_parent_p->update(delta);`